core/src/main/java/jenkins/diagnosis/BootMetrics.java

```diff
@@ -0,0 +1,149 @@
+package jenkins.diagnosis;
+
+import hudson.init.InitMilestone;
+import hudson.init.InitReactorListener;
+import jenkins.model.Jenkins;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.LineIterator;
+import org.jvnet.hudson.reactor.Milestone;
+import org.jvnet.hudson.reactor.Task;
+import org.kohsuke.MetaInfServices;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.text.MessageFormat;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Record system metrics during jenkins boot, to help diagnose performance issues.
+ * @author <a href="mailto:nicolas.deloof@gmail.com">Nicolas De Loof</a>
+ */
+@MetaInfServices
+public class BootMetrics implements InitReactorListener {
+
+    @Override
+    public void onTaskStarted(Task t) {
+    }
+
+    @Override
+    public void onTaskCompleted(Task t) {
+    }
+
+    @Override
+    public void onTaskFailed(Task t, Throwable err, boolean fatal) {
+    }
+
+    @Override
+    public void onAttained(Milestone milestone) {
+        final String name = milestone.toString();
+        if (name != null) out.println(System.currentTimeMillis() + " > " + name);
+        if (milestone == InitMilestone.COMPLETED) {
+            timer.cancel();
```

Comment:
Would an NPE here (if e.g. creation of the `logs` directory fails) fail the Jenkins startup?